File: prep/implementation/iterations/iteration_cpp/binary_tree_dynamic_node.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

template <typename T> struct TreeNode {
  T data;
  TreeNode *left;
  TreeNode *right;

  TreeNode(T data) {
    this->data = data;
    this->left = this->right = nullptr;
  }

  ~TreeNode() {
    if (this->left) {
      delete this->left;
    }

    if (this->right) {
      delete this->right;
    }
  }
};

template <typename T> class BinaryTree {
private:
  TreeNode<T> *root;

  TreeNode<T> *build(TreeNode<T> *node, int curr_idx,
                     const vector<string> &nodes, function<T(string)> parser) {
    if (curr_idx >= nodes.size() || nodes[curr_idx] == "null") {
      return nullptr;
    }

    if (node == nullptr) {
      node = new TreeNode(parser(nodes[curr_idx]));
    }

    node->left = build(node->left, 2 * curr_idx + 1, nodes, parser);
    node->right = build(node->right, 2 * curr_idx + 2, nodes, parser);

    return node;
  }

public:
  BinaryTree(const vector<string> &nodes, function<T(string)> parser) {
    this->root = nullptr;
    this->root = build(this->root, 0, nodes, parser);
  }

  ~BinaryTree() { delete this->root; }

  void dfs_pre(TreeNode<T> *node) {
    if (node == nullptr) {
      return;
    }

    cout << node->data << " ";
    dfs_pre(node->left);
    dfs_pre(node->right);
  }

  void dfs_pre() {
    dfs_pre(this->root);
    cout << "\n";
  }

  void dfs_pre_iter() {
    vector<TreeNode<T> *> st;
    st.push_back(this->root);

    while (!st.empty()) {
      TreeNode<T> *node = st.back();
      st.pop_back();
      cout << node->data << " ";

      if (node->right != nullptr) {
        st.push_back(node->right);
      }

      if (node->left != nullptr) {
        st.push_back(node->left);
      }
    }
    cout << "\n";
  }

  void dfs_in(TreeNode<T> *node) {
    if (node == nullptr) {
      return;
    }

    dfs_in(node->left);
    cout << node->data << " ";
    dfs_in(node->right);
  }

  void dfs_in() {
    dfs_in(this->root);
    cout << "\n";
  }

  void dfs_in_iter() {
    vector<TreeNode<T> *> st;
    TreeNode<T> *curr_node = this->root;

    while (curr_node != nullptr || !st.empty()) {
      while (curr_node != nullptr) {
        st.push_back(curr_node);
        curr_node = curr_node->left;
      }

      if (!st.empty()) {
        curr_node = st.back();
        st.pop_back();
        cout << curr_node->data << " ";
        curr_node = curr_node->right;
      }
    }
    cout << "\n";
  }

  void dfs_post(TreeNode<T> *node) {
    if (node == nullptr) {
      return;
    }

    dfs_post(node->left);
    dfs_post(node->right);
    cout << node->data << " ";
  }

  void dfs_post() {
    dfs_post(this->root);
    cout << "\n";
  }

  void dfs_post_iter() {
    vector<TreeNode<T> *> st;
    vector<TreeNode<T> *> rev_st;

    st.push_back(this->root);
    while (!st.empty()) {
      TreeNode<T> *node = st.back();
      st.pop_back();

      rev_st.push_back(node);

      if (node->left != nullptr) {
        st.push_back(node->left);
      }

      if (node->right != nullptr) {
        st.push_back(node->right);
      }
    }

    while (!rev_st.empty()) {
      cout << rev_st.back()->data << " ";
      rev_st.pop_back();
    }
    cout << "\n";
  }

  void bfs() {
    queue<TreeNode<T> *> q;
    q.push(this->root);

    while (!q.empty()) {
      TreeNode<T> *node = q.front();
      q.pop();

      cout << node->data << " ";

      if (node->left != nullptr) {
        q.push(node->left);
      }

      if (node->right != nullptr) {
        q.push(node->right);
      }
    }
    cout << "\n";
  }

  int height(TreeNode<T> *node) {
    if (node == nullptr) {
      return 0;
    }

    return 1 + max(height(node->left), height(node->right));
  }

  int height() { return height(this->root); }

  void bfs_rec(TreeNode<T> *node, int level) {
    if (node == nullptr) {
      return;
    }

    if (level == 0) {
      cout << node->data << " ";
    } else {
      bfs_rec(node->left, level - 1);
      bfs_rec(node->right, level - 1);
    }
  }

  void bfs_rec() {
    int h = height();
    for (int i = 0; i < h; i++) {
      bfs_rec(this->root, i);
    }
    cout << "\n";
  }
};

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  string inp;
  getline(cin, inp);

  string tmp;
  stringstream ss(inp);
  vector<string> nodes;

  while (getline(ss, tmp, ' ')) {
    nodes.push_back(tmp);
  }

  BinaryTree<int> tree(nodes, [](const string &inp) { return stoi(inp); });

  tree.dfs_pre();
  tree.dfs_pre_iter();

  tree.dfs_in();
  tree.dfs_in_iter();

  tree.dfs_post();
  tree.dfs_post_iter();

  tree.bfs();
  tree.bfs_rec();

  return 0;
}

File: prep/implementation/practice_iterations/main_1/BinaryTreeDynamicNode.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class BinaryTreeDynamicNode {
  static class BinaryTree {
    static class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;

      TreeNode(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
      }
    }

    private TreeNode root;

    private TreeNode buildTree(String[] nodes) { return buildTree(nodes, 0); }

    private TreeNode buildTree(String[] nodes, int currIdx) {
      if (currIdx >= nodes.length || nodes[currIdx].equalsIgnoreCase("null")) {
        return null;
      }

      TreeNode node = new TreeNode(Integer.parseInt(nodes[currIdx]));
      node.left = buildTree(nodes, 2 * currIdx + 1);
      node.right = buildTree(nodes, 2 * currIdx + 2);

      return node;
    }

    public BinaryTree(String[] nodes) { this.root = buildTree(nodes); }

    // O(n) / O(n)
    public void dfsPre() {
      if (root == null) {
        return;
      }

      dfsPre(root);
      System.out.println();
    }

    private void dfsPre(TreeNode node) {
      if (node == null) {
        return;
      }

      System.out.printf("%d ", node.val);

      if (node.left != null) {
        dfsPre(node.left);
      }

      if (node.right != null) {
        dfsPre(node.right);
      }
    }

    // O(n) / O(n)
    public void dfsPreIter() {
      if (root == null) {
        return;
      }

      dfsPreIter(root);
      System.out.println();
    }

    private void dfsPreIter(TreeNode node) {
      Stack<TreeNode> st = new Stack<>();
      st.push(node);

      while (!st.isEmpty()) {
        node = st.pop();

        System.out.printf("%d ", node.val);

        if (node.right != null) {
          st.push(node.right);
        }

        if (node.left != null) {
          st.push(node.left);
        }
      }
    }

    // O(n) / O(n)
    public void dfsIn() {
      if (root == null) {
        return;
      }

      dfsIn(root);
      System.out.println();
    }

    private void dfsIn(TreeNode node) {
      if (node == null) {
        return;
      }

      dfsIn(node.left);
      System.out.printf("%d ", node.val);
      dfsIn(node.right);
    }

    // O(n) / O(n)
    public void dfsInIter() {
      if (root == null) {
        return;
      }

      dfsInIter(root);
      System.out.println();
    }

    private void dfsInIter(TreeNode node) {
      Stack<TreeNode> st = new Stack<>();
      TreeNode currNode = node;

      while (currNode != null || !st.isEmpty()) {
        while (currNode != null) {
          st.push(currNode);
          currNode = currNode.left;
        }

        if (!st.isEmpty()) {
          currNode = st.pop();
          System.out.printf("%d ", currNode.val);
          currNode = currNode.right;
        }
      }
    }

    // O(n) / O(n)
    public void dfsPost() {
      if (root == null) {
        return;
      }

      dfsPost(root);
      System.out.println();
    }

    private void dfsPost(TreeNode node) {
      if (node == null) {
        return;
      }

      dfsPost(node.left);
      dfsPost(node.right);
      System.out.printf("%d ", node.val);
    }

    // O(n) / O(n)
    public void dfsPostIter() {
      if (root == null) {
        return;
      }

      dfsPostIter(root);
      System.out.println();
    }

    private void dfsPostIter(TreeNode node) {
      Stack<TreeNode> st = new Stack<>();
      Stack<TreeNode> revSt = new Stack<>();

      st.push(node);

      while (!st.isEmpty()) {
        node = st.pop();
        revSt.push(node);

        if (node.left != null) {
          st.push(node.left);
        }

        if (node.right != null) {
          st.push(node.right);
        }
      }

      while (!revSt.isEmpty()) {
        System.out.printf("%d ", revSt.pop().val);
      }
    }

    // O(n) / O(n)
    public void bfs() {
      if (root == null) {
        return;
      }

      Queue<TreeNode> q = new ArrayDeque<>();
      q.add(root);

      while (!q.isEmpty()) {
        TreeNode node = q.poll();

        System.out.printf("%d ", node.val);

        if (node.left != null) {
          q.add(node.left);
        }

        if (node.right != null) {
          q.add(node.right);
        }
      }

      System.out.println();
    }

    // O(n) / O(n)
    public void bfsRec() {
      if (root == null) {
        return;
      }

      Queue<TreeNode> q = new ArrayDeque<>();
      q.add(root);

      bfsRec(q);
      System.out.println();
    }

    private void bfsRec(Queue<TreeNode> q) {
      if (q.isEmpty()) {
        return;
      }

      TreeNode node = q.poll();
      System.out.printf("%d ", node.val);

      if (node.left != null) {
        q.add(node.left);
      }

      if (node.right != null) {
        q.add(node.right);
      }

      bfsRec(q);
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      String[] nodes = in.nextLine().split(" ");
      BinaryTree tree = new BinaryTree(nodes);

      tree.dfsPre();
      tree.dfsPreIter();

      tree.dfsIn();
      tree.dfsInIter();

      tree.dfsPost();
      tree.dfsPostIter();

      tree.bfs();
      tree.bfsRec();
    }
  }
}